app/services/GuaranteeBalanceService.scala

```diff
@@ -18,36 +18,76 @@ package services
 
 import akka.actor.ActorSystem
 import akka.pattern.after
+import config.FrontendAppConfig
 import connectors.GuaranteeBalanceConnector
-import models.backend.{BalanceRequestPending, BalanceRequestResponse}
-import models.values.BalanceId
-import uk.gov.hmrc.http.{HeaderCarrier, HttpResponse}
 import javax.inject.Inject
-import models.requests.BalanceRequest
+import models.backend.{BalanceRequestPending, BalanceRequestRateLimit, BalanceRequestResponse, BalanceRequestSessionExpired}
+import models.requests.{BalanceRequest, DataRequest}
+import models.values._
+import pages.{AccessCodePage, EoriNumberPage, GuaranteeReferenceNumberPage}
+import play.api.Logging
+import uk.gov.hmrc.http.{HeaderCarrier, HttpResponse}
+import uk.gov.hmrc.mongo.lock.MongoLockRepository
 
 import scala.concurrent.duration._
 import scala.concurrent.{ExecutionContext, Future}
 
-class GuaranteeBalanceService @Inject() (val actorSystem: ActorSystem, val connector: GuaranteeBalanceConnector)(implicit ec: ExecutionContext) {
+class GuaranteeBalanceService @Inject() (actorSystem: ActorSystem,
+                                         connector: GuaranteeBalanceConnector,
+                                         mongoLockRepository: MongoLockRepository,
+                                         config: FrontendAppConfig
+)(implicit ec: ExecutionContext)
+    extends Logging {
 
-  def submitBalanceRequest(balanceRequest: BalanceRequest)(implicit hc: HeaderCarrier): Future[Either[HttpResponse, BalanceRequestResponse]] =
-    connector.submitBalanceRequest(balanceRequest)
+  def submitBalanceRequest()(implicit hc: HeaderCarrier, request: DataRequest[_]): Future[Either[HttpResponse, BalanceRequestResponse]] =
+    (for {
+      guaranteeReferenceNumber <- request.userAnswers.get(GuaranteeReferenceNumberPage)
+      taxIdentifier            <- request.userAnswers.get(EoriNumberPage)
+      accessCode               <- request.userAnswers.get(AccessCodePage)
+    } yield checkRateLimit(request.internalId, guaranteeReferenceNumber).flatMap {
+      lockFree =>
+        if (lockFree) {
+          connector
+            .submitBalanceRequest(
+              BalanceRequest(
+                TaxIdentifier(taxIdentifier),
+                GuaranteeReference(guaranteeReferenceNumber),
+                AccessCode(accessCode)
+              )
+            )
+        } else {
+          logger.warn("[GuaranteeBalanceService][submit] Rate Limit hit")
+          Future.successful(Right(BalanceRequestRateLimit()))
+        }
+    }).getOrElse {
+      logger.warn("[GuaranteeBalanceService][submit] Insufficient data in user answers.")
+      Future.successful(Right(BalanceRequestSessionExpired()))
+    }
 
-  def pollForGuaranteeBalance(balanceId: BalanceId, delay: FiniteDuration, maxTime: FiniteDuration)(implicit
+  private def checkRateLimit(eoriNumber: String, guaranteeReferenceNumber: String): Future[Boolean] = {
```

Comment:
I spotted this yesterday but makes sense to change it here - it looks like the value we pass in is an internal ID so can we change the variable name please?

Reply:
That would definitely be a better idea. I'll change that

Reply:
@jackheslop96 do we also need to be releasing the lock here like you identified yesterday?

Reply:
@jackheslop96 I'll change the variable name. @adamconder Not sure what you mean about releasing the lock though, if we release the lock and the user can simply click Try Again and submit..... which may be what you're saying you want